$NetBSD: acroread.diff,v 1.2.2.1 2006/04/09 19:38:28 salo Exp $

--- acroread.orig	2005-11-07 12:05:17.000000000 +0000
+++ acroread
@@ -214,7 +214,7 @@ else
  ScriptDirectory="`/bin/pwd 2> /dev/null`"
  cd "$cwd"
 fi
-install_dir=`dirname "$ScriptDirectory"`/Reader
+install_dir=@PREFIX@/Acrobat7/Reader
 
 
 #
@@ -439,18 +439,18 @@ check_gtk_ver_and_set_lib_path()
     OLD_IFS=""
 
     # This checks for libs in cache /etc/ld.so.cache
-    for ifile in `/sbin/ldconfig -p|grep "libgtk-x11-2.0.so$"|sed 's/.*[ \t]*=>[ \t]*\(.*\)/\1/g' 2>/dev/null`; do
-        version=`get_gtk_file_ver "$ifile"`
-
-    	while [ ${#version} -gt ${#base_version} ]; do
-	   		base_version="${base_version}0"
-	    done
-
-        if [ $version -ge $base_version ]; then
-            export LD_LIBRARY_PATH="`dirname "$ifile"`":"$LD_LIBRARY_PATH"
-            return 0
-        fi
-    done
+#    for ifile in `/sbin/ldconfig -p|grep "libgtk-x11-2.0.so$"|sed 's/.*[ \t]*=>[ \t]*\(.*\)/\1/g' 2>/dev/null`; do
+#        version=`get_gtk_file_ver "$ifile"`
+#
+#    	while [ ${#version} -gt ${#base_version} ]; do
+#	   		base_version="${base_version}0"
+#	    done
+#
+#        if [ $version -ge $base_version ]; then
+#            export LD_LIBRARY_PATH="`dirname "$ifile"`":"$LD_LIBRARY_PATH"
+#            return 0
+#        fi
+#    done
 
     return 1
 }
@@ -577,6 +577,19 @@ case "$os_name" in
     ACRO_CONFIG=intellinux
     export ACRO_CONFIG
     ;;
+   NetBSD | DragonFly)
+     machine_type=`uname -p`
+     case "$machine_type" in
+       i386)
+         ACRO_CONFIG=intellinux
+         export ACRO_CONFIG
+         ;;
+       sparc*)
+         ACRO_CONFIG=sparcsolaris
+         export ACRO_CONFIG
+         ;;
+     esac
+     ;;
 esac
 
 if [ -z "$ACRO_CONFIG" ] ; then
@@ -611,7 +624,7 @@ case "$ACRO_CONFIG" in
     ;;
 esac
 
-DefaultPSRESPATH="$HOME/psres:$HOME/fonts:/usr/psres"
+DefaultPSRESPATH="$HOME/psres:$HOME/fonts:/usr/psres:@ACROREAD_FONTPATH@"
 if [ -z "$PSRESOURCEPATH" ] ; then
   PSRESOURCEPATH="$DefaultPSRESPATH"
 else
