$NetBSD: acroread.diff,v 1.6 2008/08/11 14:29:20 is Exp $

--- acroread.orig	2006-05-23 01:33:35.000000000 +0000
+++ acroread	2006-06-13 13:55:00.000000000 +0000
@@ -1,6 +1,12 @@
 #!/bin/sh
 #
 
+case `/usr/bin/uname` in
+NetBSD) case `/usr/bin/uname -p` in
+	i386) LD_LIBRARY_PATH=""; export LD_LIBRARY_PATH;;
+	x86_64) LD_LIBRARY_PATH=""; export LD_LIBRARY_PATH;;
+	esac;;
+esac
 
 setLangENU()
 {
@@ -233,7 +233,7 @@ else
  ScriptDirectory="`/bin/pwd 2> /dev/null`"
  cd "$cwd"
 fi
-install_dir=`dirname "$ScriptDirectory"`/Reader
+install_dir=@PREFIX@/Acrobat7/Reader
 
 
 #
@@ -476,18 +476,18 @@ check_gtk_ver_and_set_lib_path()
     OLD_IFS=""
 
     # This checks for libs in cache /etc/ld.so.cache
-    for ifile in `/sbin/ldconfig -p|grep "libgtk-x11-2.0.so$"|sed 's/.*[ \t]*=>[ \t]*\(.*\)/\1/g' 2>/dev/null`; do
-        version=`get_gtk_file_ver "$ifile"`
-
-    	while [ ${#version} -gt ${#base_version} ]; do
-	   		base_version="${base_version}0"
-	    done
-
-        if [ $version -ge $base_version ]; then
-            export LD_LIBRARY_PATH="`dirname "$ifile"`":"$LD_LIBRARY_PATH"
-            return 0
-        fi
-    done
+#    for ifile in `/sbin/ldconfig -p|grep "libgtk-x11-2.0.so$"|sed 's/.*[ \t]*=>[ \t]*\(.*\)/\1/g' 2>/dev/null`; do
+#        version=`get_gtk_file_ver "$ifile"`
+#
+#    	while [ ${#version} -gt ${#base_version} ]; do
+#	   		base_version="${base_version}0"
+#	    done
+#
+#        if [ $version -ge $base_version ]; then
+#            export LD_LIBRARY_PATH="`dirname "$ifile"`":"$LD_LIBRARY_PATH"
+#            return 0
+#        fi
+#    done
 
     return 1
 }
@@ -614,6 +614,19 @@ case "$os_name" in
     ACRO_CONFIG=intellinux
     export ACRO_CONFIG
     ;;
+  NetBSD | DragonFly)
+    machine_type=`uname -p`
+    case "$machine_type" in
+      i386|x86_64)
+        ACRO_CONFIG=intellinux
+        export ACRO_CONFIG
+        ;;
+      sparc*)
+        ACRO_CONFIG=sparcsolaris
+        export ACRO_CONFIG
+        ;;
+    esac
+    ;;
   AIX)
     ACRO_CONFIG=rs6000aix
     export ACRO_CONFIG
@@ -668,7 +681,7 @@ case "$ACRO_CONFIG" in
     ;;
 esac
 
-DefaultPSRESPATH="$HOME/psres:$HOME/fonts:/usr/psres"
+DefaultPSRESPATH="$HOME/psres:$HOME/fonts:/usr/psres:@ACROREAD_FONTPATH@"
 if [ -z "$PSRESOURCEPATH" ] ; then
   PSRESOURCEPATH="$DefaultPSRESPATH"
 else
