$NetBSD: info.patch,v 1.1.1.1 2000/11/21 05:21:08 kei Exp $

--- info/CCL.orig	Thu Nov 16 23:13:27 2000
+++ info/CCL	Thu Nov 16 23:16:47 2000
@@ -3,7 +3,10 @@
 from file `CCL.texi'
 using `texinfmt.el' version 2.32 of 19 November 1993.
 
-
+INFO-DIR-SECTION Editors
+START-INFO-DIR-ENTRY
+* CCL: (CCL).		Code Conversion Language.
+END-INFO-DIR-ENTRY
 
 
 File: CCL, Node: Top, Up: (mule)
--- info/ISO2022.orig	Thu Nov 16 23:14:20 2000
+++ info/ISO2022	Thu Nov 16 23:16:52 2000
@@ -3,7 +3,10 @@
 from file `ISO2022.texi'
 using `texinfmt.el' version 2.32 of 19 November 1993.
 
-
+INFO-DIR-SECTION Editors
+START-INFO-DIR-ENTRY
+* ISO2022: (ISO2022).	ISO2022 encoding system.
+END-INFO-DIR-ENTRY
 
 
 File: ISO2022, Node: Top, Up: (mule)
--- info/R2L.orig	Thu Nov 16 23:16:04 2000
+++ info/R2L	Thu Nov 16 23:16:32 2000
@@ -3,7 +3,10 @@
 from file `R2L.texi'
 using `texinfmt.el' version 2.32 of 19 November 1993.
 
-
+INFO-DIR-SECTION Editors
+START-INFO-DIR-ENTRY
+* R2L: (R2L).		Right-to-left writing on Mule.
+END-INFO-DIR-ENTRY
 
 
 File: R2L, Node: Top, Up: (mule)
--- info/XFONT.orig	Thu Nov 16 23:35:33 2000
+++ info/XFONT	Thu Nov 16 23:37:17 2000
@@ -2,7 +2,9 @@
 produced by texinfo-format-buffer
 from file: XFONT.texi
 
-
+START-INFO-DIR-ENTRY
+* XFONT: (XFONT).	X's FONT usage.
+END-INFO-DIR-ENTRY
 
 
 File: XFONT  Node: Top, Up: (mule)
--- info/antenews-jp.orig	Thu Nov 16 23:57:34 2000
+++ info/antenews-jp	Thu Nov 16 23:59:26 2000
@@ -3,6 +3,10 @@
 from file `antenews-jp.texi'
 using `texinfmt.el' version 2.32 of 19 November 1993.
 
+INFO-DIR-SECTION Editors
+START-INFO-DIR-ENTRY
+* Antenews-jp: (antenews-jp). Antenews of Emacs 19. (written in Japanese)
+END-INFO-DIR-ENTRY
 
 
 File: antenews-jp, Node: Top, Next: Basic Changes, Up: (dir)
--- info/canna-jp.orig	Thu Nov 16 23:17:23 2000
+++ info/canna-jp	Thu Nov 16 23:27:01 2000
@@ -2,7 +2,11 @@
 produced by texinfo-format-buffer
 from file: canna-jp.texinfo
 
-
+INFO-DIR-SECTION Editors
+START-INFO-DIR-ENTRY
+* Canna-jp: (canna-jp).	Japanese Inputting method Canna.
+			  (written in Japanese)
+END-INFO-DIR-ENTRY
 
 
 File: canna-jp  Node: Top, Prev: (dir), Up: (dir), Next: 概説
--- info/egg-jp-1.orig	Thu Nov 16 23:24:54 2000
+++ info/egg-jp-1	Thu Nov 16 23:26:29 2000
@@ -2,8 +2,11 @@
 produced by texinfo-format-buffer
 from file: egg-jp.texi
 
-
-
+INFO-DIR-SECTION Editors
+START-INFO-DIR-ENTRY
+* Egg-jp: (egg-jp).	Built-in Chinese/Japanese Inputting Method for Mule.
+			  (written in Japanese)
+END-INFO-DIR-ENTRY
 
 
 File: egg-jp  Node: Top, Up: (dir), Next: Overview
--- info/egg-jp-2.orig	Thu Nov 16 23:25:00 2000
+++ info/egg-jp-2	Thu Nov 16 23:26:40 2000
@@ -2,8 +2,11 @@
 produced by texinfo-format-buffer
 from file: egg-jp.texi
 
-
-
+INFO-DIR-SECTION Editors
+START-INFO-DIR-ENTRY
+* Egg-jp: (egg-jp).	Built-in Chinese/Japanese Inputting Method for Mule.
+			  (written in Japanese)
+END-INFO-DIR-ENTRY
 
 
 File: egg-jp  Node: Editing, Prev: Registration, Up: Dictionary Management, Next: Customization
--- info/egg-jp.orig	Thu Nov 16 23:24:46 2000
+++ info/egg-jp	Thu Nov 16 23:26:20 2000
@@ -2,8 +2,11 @@
 produced by texinfo-format-buffer
 from file: egg-jp.texi
 
-
-
+INFO-DIR-SECTION Editors
+START-INFO-DIR-ENTRY
+* Egg-jp: (egg-jp).	Built-in Chinese/Japanese Inputting Method for Mule.
+			  (written in Japanese)
+END-INFO-DIR-ENTRY
 
 
 Indirect:
--- info/egg.orig	Thu Nov 16 23:22:57 2000
+++ info/egg	Thu Nov 16 23:28:48 2000
@@ -3,7 +3,10 @@
 from file `egg.texi'
 using `texinfmt.el' version 2.32 of 19 November 1993.
 
-
+INFO-DIR-SECTION Editors
+START-INFO-DIR-ENTRY
+* egg: (egg).		Built-in Chinese/Japanese Inputting method for Mule.
+END-INFO-DIR-ENTRY
 
 
 File: egg, Node: Top, Up: (mule)
--- info/kbd-trans.orig	Thu Nov 16 23:37:34 2000
+++ info/kbd-trans	Thu Nov 16 23:38:39 2000
@@ -3,7 +3,10 @@
 from file `kbd-trans.texi'
 using `texinfmt.el' version 2.32 of 19 November 1993.
 
-
+INFO-DIR-SECTION Editors
+START-INFO-DIR-ENTRY
+* kbd-trans: (kbd-trans). keyboard translation on Mule.
+END-INFO-DIR-ENTRY
 
 
 File: kbd-trans, Node: Top, Up: (mule)
--- info/languages.orig	Thu Nov 16 23:39:23 2000
+++ info/languages	Thu Nov 16 23:41:03 2000
@@ -3,7 +3,10 @@
 from file `languages.texi'
 using `texinfmt.el' version 2.32 of 19 November 1993.
 
-
+INFO-DIR-SECTION Editors
+START-INFO-DIR-ENTRY
+* languages: (languages). various Asian and European languages in Mule.
+END-INFO-DIR-ENTRY
 
 
 File: languages, Node: Top, Up: (mule)
--- info/m2ps.orig	Thu Nov 16 23:41:11 2000
+++ info/m2ps	Thu Nov 16 23:41:46 2000
@@ -3,7 +3,10 @@
 from file `m2ps.texi'
 using `texinfmt.el' version 2.32 of 19 November 1993.
 
-
+INFO-DIR-SECTION Editors
+START-INFO-DIR-ENTRY
+* m2ps: (m2ps).		Converter from Mule code to PostScript.
+END-INFO-DIR-ENTRY
 
 
 File: m2ps, Node: Top, Up: (mule)
--- info/mule-jp-1.orig	Thu Nov 16 23:53:22 2000
+++ info/mule-jp-1	Thu Nov 16 23:54:40 2000
@@ -3,7 +3,11 @@
 from file `mule-jp.texi'
 using `texinfmt.el' version 2.32 of 19 November 1993.
 
-
+INFO-DIR-SECTION Editors
+START-INFO-DIR-ENTRY
+* Mule-jp: (mule-jp).		MULtilingual Enhancement to GNU Emacs.
+			  (written in Japanese)
+END-INFO-DIR-ENTRY
 
 
 File: mule-jp, Node: Top, Next: Overview, Up: (dir)
--- info/mule-jp-2.orig	Thu Nov 16 23:53:22 2000
+++ info/mule-jp-2	Thu Nov 16 23:54:42 2000
@@ -3,7 +3,11 @@
 from file `mule-jp.texi'
 using `texinfmt.el' version 2.32 of 19 November 1993.
 
-
+INFO-DIR-SECTION Editors
+START-INFO-DIR-ENTRY
+* Mule-jp: (mule-jp).		MULtilingual Enhancement to GNU Emacs.
+			  (written in Japanese)
+END-INFO-DIR-ENTRY
 
 
 File: mule-jp, Node: Set Coding System, Next: Key Input, Prev: Buffers, Up: Multilingual Char
--- info/mule-jp-3.orig	Thu Nov 16 23:53:22 2000
+++ info/mule-jp-3	Thu Nov 16 23:54:45 2000
@@ -3,7 +3,11 @@
 from file `mule-jp.texi'
 using `texinfmt.el' version 2.32 of 19 November 1993.
 
-
+INFO-DIR-SECTION Editors
+START-INFO-DIR-ENTRY
+* Mule-jp: (mule-jp).		MULtilingual Enhancement to GNU Emacs.
+			  (written in Japanese)
+END-INFO-DIR-ENTRY
 
 
 File: mule-jp, Node: Justifying, Next: Enhancement, Prev: Filling, Up: Filling&Justifying
--- info/mule-jp.orig	Thu Nov 16 23:53:22 2000
+++ info/mule-jp	Thu Nov 16 23:54:36 2000
@@ -3,7 +3,11 @@
 from file `mule-jp.texi'
 using `texinfmt.el' version 2.32 of 19 November 1993.
 
-
+INFO-DIR-SECTION Editors
+START-INFO-DIR-ENTRY
+* Mule-jp: (mule-jp).		MULtilingual Enhancement to GNU Emacs.
+			  (written in Japanese)
+END-INFO-DIR-ENTRY
 
 
 Indirect:
--- info/mule.orig	Thu Nov 16 23:46:33 2000
+++ info/mule	Thu Nov 16 23:51:31 2000
@@ -3,6 +3,10 @@
 from file `mule.texi'
 using `texinfmt.el' version 2.32 of 19 November 1993.
 
+INFO-DIR-SECTION Editors
+START-INFO-DIR-ENTRY
+* Mule: (mule).		MULtilingual Enhancement to GNU Emacs.
+END-INFO-DIR-ENTRY
 
 This file documents the internal structure of Mule (MULtilingual
 Enhancement to GNU Emacs).
--- info/quail.orig	Thu Nov 16 23:53:27 2000
+++ info/quail	Thu Nov 16 23:55:28 2000
@@ -3,7 +3,10 @@
 from file `quail.texi'
 using `texinfmt.el' version 2.32 of 19 November 1993.
 
-
+INFO-DIR-SECTION Editors
+START-INFO-DIR-ENTRY
+* quail: (quail).	Imputting methods of multilingual text on Mule.
+END-INFO-DIR-ENTRY
 
 
 File: quail, Node: Top, Up: (mule)
--- info/terminology.orig	Thu Nov 16 23:59:54 2000
+++ info/terminology	Fri Nov 17 00:01:42 2000
@@ -3,7 +3,10 @@
 from file `terminology.texi'
 using `texinfmt.el' version 2.32 of 19 November 1993.
 
-
+INFO-DIR-SECTION Editors
+START-INFO-DIR-ENTRY
+* Terminology: (terminology). Terminology used by Mule.
+END-INFO-DIR-ENTRY
 
 
 File: terminology, Node: Top, Up: (mule)
