Make all ObjC methods saveds.

--- gcc/objc/objc-act.cc.orig	2025-08-08 06:51:41.003356964 +0000
+++ gcc/objc/objc-act.cc
@@ -5221,6 +5221,13 @@ build_function_type_for_method (tree ret
   else
     ftype = build_function_type_vec (return_type, argtypes);
 
+#ifdef TARGET_BASEREL
+  if (TARGET_BASEREL)
+    ftype =
+      build_type_attribute_variant (ftype, tree_cons (get_identifier ("saveds"),
+                                    NULL_TREE, TYPE_ATTRIBUTES (ftype)));
+#endif
+
   release_tree_vector (argtypes);
   return ftype;
 }
