$NetBSD: irix5.patch,v 1.3 2006/09/10 13:44:58 schwarz Exp $

--- sgi.cf.orig	2006-09-09 17:13:25.000000000 +0200
+++ sgi.cf	2006-09-10 10:15:18.000000000 +0200
@@ -178,8 +178,10 @@
 #define AsCmd $(TOOLROOT)/usr/bin/as
 #define CcCmd $(TOOLROOT)/usr/bin/cc
 #ifdef UseInstalled
+#ifndef InstallCmd
 #define InstallCmd bsdinst
 #endif
+#endif
 #ifndef LdCmd
 #define LdCmd $(TOOLROOT)/usr/bin/ld
 #endif
@@ -188,12 +190,14 @@
  * Actually, it doesn't matter what MakeCmd is, since we don't use it.
  * We assume that $(MAKE) gets set for us by make.
  */
-#define MakeCmd $(TOOLROOT)/bin/make
+/*#define MakeCmd $(TOOLROOT)/bin/make */
 #define TroffCmd psroff -t
 
 #define StandardIncludes -nostdinc -I$(ROOT)/usr/include
 
+#ifndef ExtraLibraries
 #define ExtraLibraries /**/
+#endif
 
 /*
  * If we specify -nostdlib, we must supply the complete library
@@ -252,23 +256,19 @@
 #define ManSuffix	1
 #define LibManSuffix	3
 
-#undef ManDir
-#undef LibmanDir
-#define ManDir $(CATMAN1DIR)$(MANPACKAGE)
-#define LibmanDir $(CATMAN3DIR)$(MANPACKAGE)
 #define NroffManDir $(MAN1DIR)$(MANPACKAGE)
 #define NroffLibmanDir $(MAN3DIR)$(MANPACKAGE)
 
 #ifdef UseInstalled
-           MANUSR = /usr/share/local
+           MANUSR = ManUsr
 #else
            MANUSR = /usr/share
 #endif
 
-       CATMAN1DIR = $(MANUSR)/catman/u_man/cat1
-       CATMAN3DIR = $(MANUSR)/catman/p_man/cat3
-          MAN1DIR = $(MANUSR)/man/u_man/man1
-          MAN3DIR = $(MANUSR)/man/p_man/man3
+       CATMAN1DIR = $(MANUSR)/tman/cat1
+       CATMAN3DIR = $(MANUSR)/tman/cat3
+          MAN1DIR = $(MANUSR)/man/man1
+          MAN3DIR = $(MANUSR)/man/man3
 
 /*
  * MANPACKAGE is the directory inside of cat? or man? that man pages should
@@ -281,22 +281,22 @@
  * InstallManPageLong - misc changes for sgi.
  */
 #define InstallManPageLong(file,destdir,dest)				@@\
-file.nr.z : file.man							@@\
+file.nr.gz : file.man							@@\
 	@if [ -f file.man ]; then \					@@\
-		$(RM) file.nr file.nr.z && \				@@\
+		$(RM) file.nr file.nr.gz && \				@@\
 		tbl file.man | eqn | nroff -man - > file.nr && \	@@\
-		pack -f file.nr; \					@@\
+		$(GZIP_CMD) file.nr; \					@@\
 	fi								@@\
 									@@\
-install.man:: file.nr.z							@@\
+install.man:: file.nr.gz						@@\
 	MakeDir($(DESTDIR)destdir); \					@@\
 	if [ -f file.man ]; then \					@@\
 		$(INSTALL) -c $(INSTMANFLAGS) \				@@\
-			file.nr.z $(DESTDIR)destdir/dest.z; \		@@\
+			file.nr.gz $(DESTDIR)destdir/dest.$(MANSUFFIX).gz; \	@@\
 	fi								@@\
 									@@\
 clean::									@@\
-	$(RM) file.nr file.nr.z
+	$(RM) file.nr file.nr.gz
 
 /*
  * InstallManPageAliases - generate rules to install manual page aliases.
--- sgiLib.rules.orig	2006-09-10 10:29:23.000000000 +0200
+++ sgiLib.rules	2006-09-09 21:06:23.000000000 +0200
@@ -97,8 +97,8 @@
 	$(INSTALL) -c $(INSTLIBFLAGS) Concat4(lib,libname,.so,rev) \	@@\
 		$(DESTDIR)dest ; \					@@\
 	STR=rev; if [ -n "$${STR}" ]; then \				@@\
-	    $(INSTALL) -c -l Concat4(lib,libname,.so,rev)  \		@@\
-		Concat3(lib,libname,.so) $(DESTDIR)dest ; \		@@\
+	    $(LN) Concat4(lib,libname,.so,rev)  \			@@\
+		$(DESTDIR)dest/Concat3(lib,libname,.so) ; \		@@\
 	fi
 
 /*
