$NetBSD: irix6.patch,v 1.1 2006/04/10 13:38:36 schwarz Exp $

--- X11.rules.orig	Thu Jun 12 09:12:20 2003
+++ X11.rules	Sun Jun  8 10:31:35 2003
@@ -79,7 +79,7 @@
 #if UseInstalledX11 
 #if defined(X11ProjectRoot)
 # define X11BuildLibPath	$(XPROJECTROOT)/lib
-else
+#else
 # define X11BuildLibPath        $(USRLIBDIR)
 #endif
 #elif ImportX11
--- sgi.cf.orig	Thu Jun 12 09:12:20 2003
+++ sgi.cf		Sun Jun  8 12:55:50 2003
@@ -241,7 +241,9 @@
  * Actually, it doesn't matter what MakeCmd is, since we don't use it.
  * We assume that $(MAKE) gets set for us by make.
  */
+#ifndef MakeCmd
 #define MakeCmd $(TOOLROOTSAFE) $(TOOLROOT)/sbin/make
+#endif
 #define TroffCmd psroff -t
 
 #define StandardIncludes -nostdinc -I$(ROOT)/usr/include
@@ -320,10 +322,14 @@
 #define NroffFilemanDir $(MAN4DIR)$(MANPACKAGE)
 #define NroffMiscmanDir $(MAN5DIR)$(MANPACKAGE)
 
+#ifndef ManUsr
 #ifdef UseInstalled
            MANUSR = /usr/share/local
 #else
            MANUSR = /usr/share
+#endif
+#else
+           MANUSR = ManUsr
 #endif
 
        CATMAN1DIR = $(MANUSR)/catman/u_man/cat1
