$NetBSD: ocaml-configure.diff,v 1.1.2.2 2012/04/15 14:35:20 tron Exp $

--- configure~	2011-07-04 21:15:01.000000000 +0000
+++ configure
@@ -621,7 +621,7 @@ if test $withsharedlibs = "yes"; then
     i[3456]86-*-linux*)           natdynlink=true;;
     x86_64-*-linux*)              natdynlink=true;;
     i[3456]86-*-darwin10.*)
-      if test $arch64 == true; then
+      if test $arch64 = true; then
         natdynlink=true
       fi;;
     i[3456]86-*-darwin[89]*)      natdynlink=true;;
