$NetBSD: acroread.diff,v 1.2 2004/07/24 14:43:12 mrg Exp $

--- acroread.orig	Tue Jul  2 08:08:00 2002
+++ acroread	Thu Aug  1 21:50:23 2002
@@ -191,6 +191,19 @@
     ACRO_CONFIG=intellinux
     export ACRO_CONFIG
     ;;
+   NetBSD)
+     machine_type=`uname -p`
+     case "$machine_type" in
+       i386)
+         ACRO_CONFIG=intellinux
+         export ACRO_CONFIG
+         ;;
+       sparc*)
+         ACRO_CONFIG=sparcsolaris
+         export ACRO_CONFIG
+         ;;
+     esac
+     ;;
 esac
 
 if [ -z "$ACRO_CONFIG" ] ; then
