diff -ru wxWidgets-3.0.2.old/build/bakefiles/common.bkl wxWidgets-3.0.2/build/bakefiles/common.bkl
--- wxWidgets-3.0.2.old/build/bakefiles/common.bkl	2016-03-24 13:27:07.034451400 +0000
+++ wxWidgets-3.0.2/build/bakefiles/common.bkl	2016-03-24 13:27:18.295690311 +0000
@@ -708,7 +708,6 @@
         <if cond="FORMAT=='autoconf'">
             <res-include>$(RCDEFDIR)</res-include>
             <res-include>$(TOP_SRCDIR)include</res-include>
-            <postlink-command>$(DYLIB_RPATH_POSTLINK)</postlink-command>
         </if>
         <win32-res>$(WXTOPDIR)src/msw/version.rc</win32-res>
 
diff -ru wxWidgets-3.0.2.old/build/bakefiles/config.bkl wxWidgets-3.0.2/build/bakefiles/config.bkl
--- wxWidgets-3.0.2.old/build/bakefiles/config.bkl	2016-03-24 13:27:07.034623707 +0000
+++ wxWidgets-3.0.2/build/bakefiles/config.bkl	2016-03-24 13:27:18.296182090 +0000
@@ -425,8 +425,6 @@
         <option name="EXTRALIBS_SDL"/>
         <option name="CXXWARNINGS"/>
         <option name="HOST_SUFFIX"/>
-        <option name="DYLIB_RPATH_INSTALL"/>
-        <option name="DYLIB_RPATH_POSTLINK"/>
         <option name="SAMPLES_RPATH_FLAG"/>
 
         <!-- see configure.in; it's required by some samples on Mac OS X -->
diff -ru wxWidgets-3.0.2.old/build/bakefiles/regex.bkl wxWidgets-3.0.2/build/bakefiles/regex.bkl
--- wxWidgets-3.0.2.old/build/bakefiles/regex.bkl	2016-03-24 13:27:07.035148244 +0000
+++ wxWidgets-3.0.2/build/bakefiles/regex.bkl	2016-03-24 13:27:18.296378774 +0000
@@ -26,16 +26,13 @@
         <if cond="wxUSE_REGEX=='builtin'">$(TOP_SRCDIR)src/regex</if>
     </set>
 
-    <lib id="wxregex" template="msvc_setup_h,3rdparty_lib"
+    <dll id="wxregex" template="msvc_setup_h,anylib"
          cond="wxUSE_REGEX=='builtin' and BUILDING_LIB=='1'">
-        <libname cond="FORMAT=='autoconf'">
-            $(id)$(WXUNICODEFLAG)$(WXDEBUGFLAG)$(WX_LIB_FLAVOUR)-$(WX_RELEASE)$(HOST_SUFFIX)
-        </libname>
-        <libname cond="FORMAT!='autoconf'">
-            $(id)$(WXUNICODEFLAG)$(WXDEBUGFLAG)$(HOST_SUFFIX)
-        </libname>
-        <include cond="FORMAT!='autoconf'">$(TOP_SRCDIR)include</include>
-        <include cond="FORMAT!='autoconf'">$(SETUPHDIR)</include>
+        <dllname>
+            $(id)$(WXUNICODEFLAG)$(WXDEBUGFLAG)$(WX_LIB_FLAVOUR)-$(WX_RELEASE)
+	</dllname>
+        <version>$(WX_VERSION)</version>
+        <so_version>$(WXSOVERSION)</so_version>
         <define>__WX$(TOOLKIT)__</define>
         <define>$(WXUNIV_DEFINE)</define>
         <define>$(UNICODE_DEFINE)</define>
@@ -48,6 +45,6 @@
             src/regex/regerror.c
             src/regex/regfree.c
         </sources>
-    </lib>
+    </dll>
 
 </makefile>
diff -ru wxWidgets-3.0.2.old/build/bakefiles/scintilla.bkl wxWidgets-3.0.2/build/bakefiles/scintilla.bkl
--- wxWidgets-3.0.2.old/build/bakefiles/scintilla.bkl	2016-03-24 13:27:07.035704210 +0000
+++ wxWidgets-3.0.2/build/bakefiles/scintilla.bkl	2016-03-24 13:29:43.464485658 +0000
@@ -43,12 +43,17 @@
         <if cond="SHARED=='1' and MONOLITHIC=='0'">WXUSINGDLL</if>
     </set>
 
-    <lib id="wxscintilla" template="3rdparty_lib,wxscintilla_cppflags,msvc_setup_h"
+    <dll id="wxscintilla" template="3rdparty_lib,wxscintilla_cppflags,msvc_setup_h"
          cond="USE_STC=='1' and BUILDING_LIB=='1'">
         <if cond="FORMAT!='autoconf'">
             <include>$(SETUPHDIR)</include>
             <include>$(TOP_SRCDIR)include</include>
         </if>
+        <dllname>
+            $(id)$(WXUNICODEFLAG)$(WXDEBUGFLAG)$(WX_LIB_FLAVOUR)-$(WX_RELEASE)
+	</dllname>
+        <version>$(WX_VERSION)</version>
+        <so_version>$(WXSOVERSION)</so_version>
         <define>$(wxscintilla_usingdll)</define>
         <define>__WX$(TOOLKIT)__</define>
         <define>$(WXUNIV_DEFINE)</define>
@@ -176,7 +181,7 @@
             src/stc/scintilla/src/ViewStyle.cxx
             src/stc/scintilla/src/XPM.cxx
         </sources>
-    </lib>
+    </dll>
 
     <!-- used to conditionally link against wxscintilla only if it's enabled -->
     <set var="wxscintilla_library_link">
diff -ru wxWidgets-3.0.2.old/build/bakefiles/wx.bkl wxWidgets-3.0.2/build/bakefiles/wx.bkl
--- wxWidgets-3.0.2.old/build/bakefiles/wx.bkl	2016-03-24 13:27:07.034706125 +0000
+++ wxWidgets-3.0.2/build/bakefiles/wx.bkl	2016-03-24 13:27:18.296670656 +0000
@@ -95,17 +95,11 @@
             <command>
                 $(INSTALL_DIR) $(DESTDIR)$(BINDIR)
                 $(INSTALL_DIR) $(DESTDIR)$(LIBDIR)/wx/config
-                $(INSTALL_PROGRAM) lib/wx/config/$(TOOLCHAIN_FULLNAME) $(DESTDIR)$(LIBDIR)/wx/config
+                ${BSD_INSTALL_SCRIPT} lib/wx/config/$(TOOLCHAIN_FULLNAME) $(DESTDIR)$(LIBDIR)/wx/config
                 (cd $(DESTDIR)$(BINDIR) &amp;&amp; rm -f wx-config &amp;&amp; $(LN_S) $(LIBDIR)/wx/config/$(TOOLCHAIN_FULLNAME) wx-config || cp -p $(DESTDIR)$(LIBDIR)/wx/config/$(TOOLCHAIN_FULLNAME) wx-config)
             </command>
         </action>
 
-        <modify-target target="install">
-            <command>
-                $(DYLIB_RPATH_INSTALL)
-            </command>
-        </modify-target>
-
         <set var="RCDEFS_H">
             <if cond="TOOLKIT=='MSW'">msw/rcdefs.h</if>
         </set>
diff -ru wxWidgets-3.0.2.old/utils/wxrc/wxrc.bkl wxWidgets-3.0.2/utils/wxrc/wxrc.bkl
--- wxWidgets-3.0.2.old/utils/wxrc/wxrc.bkl	2016-03-24 13:27:06.388358531 +0000
+++ wxWidgets-3.0.2/utils/wxrc/wxrc.bkl	2016-03-24 13:27:18.296926708 +0000
@@ -21,7 +21,7 @@
         <modify-target target="install_wxrc">
             <command>
 	        rm -f $(DESTDIR)$(BINDIR)/wxrc$(EXEEXT) $(DESTDIR)$(BINDIR)/$(versioned_name)
-		$(INSTALL_PROGRAM) wxrc$(EXEEXT) $(DESTDIR)$(BINDIR)
+		${LIBTOOL} --mode=install $(INSTALL_PROGRAM) wxrc$(EXEEXT) $(DESTDIR)$(BINDIR)
                 mv -f $(DESTDIR)$(BINDIR)/wxrc$(EXEEXT) $(DESTDIR)$(BINDIR)/$(versioned_name)
                 (cd $(DESTDIR)$(BINDIR) &amp;&amp; $(LN_S) $(versioned_name) wxrc$(EXEEXT))
             </command>
