$NetBSD: otps.Makefile.patch,v 1.1.1.1 2005/11/17 22:27:05 minskim Exp $

--- Makefile.orig	2005-11-17 12:14:16.000000000 -0800
+++ Makefile
@@ -73,7 +73,7 @@ LDFLAGS = -L/Users/msk/pkg/lib $(XLDFLAG
 # proglib is for web2c; 
 # XLOADLIBES is for the installer.
 LIBS = -lm 
-LOADLIBES = $(proglib) $(kpathsea) $(LIBS) $(XLOADLIBES)
+LOADLIBES = $(proglib) -lkpathsea $(LIBS) $(XLOADLIBES)
 
 # May as well separate linking from compiling, just in case.
 CCLD = $(CC)
@@ -212,7 +212,7 @@ $(LEX_OUTPUT_ROOT).c: otp.l
 outocp: outocp.o
 	$(kpathsea_link) outocp.o $(LOADLIBES)
 
-$(programs): $(proglib) $(kpathsea)
+$(programs): $(proglib)
 
 install-programs: $(programs)
 	$(SHELL) $(top_srcdir)/../mkinstalldirs $(bindir) 
