$NetBSD: acroread.diff,v 1.4 2000/04/18 01:46:17 dmcmahill Exp $

--- acroread.orig	Mon Apr 17 16:30:05 2000
+++ acroread	Mon Apr 17 16:33:30 2000
@@ -210,4 +210,21 @@
     export ACRO_CONFIG
     ;;
+  NetBSD)
+    machine_type=`uname -p`
+    case "$machine_type" in
+      alpha)
+        ACRO_CONFIG=alphaosf
+        export ACRO_CONFIG
+        ;;
+      i386)
+        ACRO_CONFIG=intellinux
+        export ACRO_CONFIG
+        ;;
+      sparc)
+        ACRO_CONFIG=sparcsolaris
+        export ACRO_CONFIG
+        ;;
+    esac
+    ;;
 esac
 
@@ -310,4 +327,6 @@
     LD_LIBRARY_PATH="`prepend "$ACRO_INSTALL_DIR/$ACRO_CONFIG/lib:$ACRO_INSTALL_DIR/$ACRO_CONFIG/lib" "$LD_LIBRARY_PATH"`"
     export LD_LIBRARY_PATH
+    LC_CTYPE="C"
+    export LC_CTYPE
     ;;
   intelbsdi)
