--- libavformat/Makefile.orig	2004-08-14 14:51:47.000000000 +0200
+++ libavformat/Makefile	2004-08-14 14:52:43.000000000 +0200
@@ -32,7 +32,7 @@
 OBJS+= framehook.o 
 
 ifeq ($(CONFIG_VIDEO4LINUX),yes)
-OBJS+= grab.o
+OBJS+= grab_bsdbktr.o
 endif
 
 ifeq ($(CONFIG_DV1394),yes)
--- configure.org       2004-07-31 17:12:40.000000000 +0200
+++ configure   2004-07-31 17:12:54.000000000 +0200
@@ -231,6 +231,8 @@
 *) ;;
 esac

+v4l="yes"
+
 # From mplayer configure. We need TARGET_OS available
 # to the Makefile, so it can distinguish between flavors
 # of AltiVec on PowerPC
