--- libavformat/Makefile.orig
+++ libavformat/Makefile
@@ -40,7 +40,7 @@
 OBJS+= framehook.o 
 
 ifeq ($(CONFIG_VIDEO4LINUX),yes)
-OBJS+= grab.o
+OBJS+= grab_bsdbktr.o
 endif
 
 ifeq ($(CONFIG_DV1394),yes)

--- configure.orig
+++ configure
@@ -301,6 +301,8 @@
 *) ;;
 esac
 
+v4l="yes"
+
 # From mplayer configure. We need TARGET_OS available
 # to the Makefile, so it can distinguish between flavors
 # of AltiVec on PowerPC
