--- Makefile.orig	2015-10-30 11:52:16.000000000 +0000
+++ Makefile	2015-10-30 11:52:35.000000000 +0000
--- Makefile.orig	2015-11-03 10:22:06.000000000 +0000
+++ Makefile
@@ -801,9 +801,6 @@ DOCUMENTATION = $(filter-out $(EXCLUDE_D
 
 install-doc:
 	install -d "$(DESTDIR)$(DOCDIR)"
-	for x in $(DOCUMENTATION); do \
-		install -m 644 doc/$$x.txt "$(DESTDIR)$(DOCDIR)" ; \
-	done
 
 install-bin: haproxy $(EXTRA)
 	install -d "$(DESTDIR)$(SBINDIR)"
