$NetBSD: acroread.diff,v 1.1.1.1 2008/04/06 01:43:32 toshii Exp $

--- acroread.orig	2008-01-16 08:47:52.000000000 +0900
+++ acroread	2008-03-29 20:01:20.000000000 +0900
@@ -17,6 +17,8 @@
 {
     if [ "`uname -s`" = "Linux" ] && [ ! -x /lib/ld-lsb.so.3 ]; then
         exec /lib/ld-linux.so.2 ${1+"$@"}
+    elif [ "$ACRO_CONFIG" = intellinux ] && [ ! -x /emul/linux/lib/ld-lsb.so.3 ]; then
+        exec /emul/linux/lib/ld-linux.so.2 ${1+"$@"}
     else
         exec ${1+"$@"}
     fi
@@ -372,6 +374,20 @@
   HP-UX)
     ACRO_CONFIG=hppahpux
     export ACRO_CONFIG
+    ;;
+  NetBSD | DragonFly)
+    machine_type=`uname -p`
+    case "$machine_type" in
+      i386|x86_64)
+        ACRO_CONFIG=intellinux
+        export ACRO_CONFIG
+        ;;
+      sparc*)
+        ACRO_CONFIG=sparcsolaris
+        export ACRO_CONFIG
+        ;;
+    esac
+    ;;
 esac
 
 if [ -z "$ACRO_CONFIG" ] ; then
