--- acroread.orig	2009-10-09 00:11:14.000000000 -0500
+++ acroread	2009-11-04 17:20:18.588874100 -0600
@@ -17,6 +17,8 @@
 {
     if [ "`uname -s`" = "Linux" ] && [ ! -x /lib/ld-lsb.so.3 ]; then
         exec /lib/ld-linux.so.2 ${1+"$@"}
+    elif [ "$ACRO_CONFIG" = intellinux ] && [ ! -x @EMULDIR@/lib/ld-lsb.so.3 ]; then
+        exec @EMULDIR@/lib/ld-linux.so.2 ${1+"$@"}
     else
         exec ${1+"$@"}
     fi
@@ -383,6 +385,20 @@
   HP-UX)
     ACRO_CONFIG=hppahpux
     export ACRO_CONFIG
+    ;;
+  NetBSD | DragonFly)
+    machine_type=`uname -p`
+    case "$machine_type" in
+      i386|x86_64)
+        ACRO_CONFIG=intellinux
+        export ACRO_CONFIG
+        ;;
+      sparc*)
+        ACRO_CONFIG=sparcsolaris
+        export ACRO_CONFIG
+        ;;
+    esac
+    ;;
 esac
 
 if [ -z "$ACRO_CONFIG" ] ; then
@@ -447,6 +463,8 @@
 case "$ACRO_CONFIG" in
   sparcsolaris)
     LD_LIBRARY_PATH="`prepend "$ACRO_INSTALL_DIR/$ACRO_CONFIG/lib:$ACRO_INSTALL_DIR/$ACRO_CONFIG/sidecars:${MOZILLA_COMP_PATH}" "${LD_LIBRARY_PATH}"`"
+    LD_NOVERSION=1
+    export LD_NOVERSION
     export LD_LIBRARY_PATH
     ;;
   intellinux)
